src/main/java/com/google/cloud/teleport/io/WindowedFilenamePolicy.java

```diff
@@ -137,22 +370,69 @@ public ResourceId unwindowedFilename(
       ValueProvider<String> outputDirectoryStr, BoundedWindow window) {
     ResourceId outputDirectory = FileSystems.matchNewResource(outputDirectoryStr.get(), true);
-
+  
     if (window instanceof IntervalWindow) {
       IntervalWindow intervalWindow = (IntervalWindow) window;
       DateTime time = intervalWindow.end().toDateTime();
       String outputPath = outputDirectory.toString();
-      outputPath = outputPath.replace("YYYY", YEAR.print(time));
-      outputPath = outputPath.replace("MM", MONTH.print(time));
-      outputPath = outputPath.replace("DD", DAY.print(time));
-      outputPath = outputPath.replace("HH", HOUR.print(time));
-      outputPath = outputPath.replace("mm", MINUTE.print(time));
+
+      // Default Date patterns.
+      String yearPattern = "YYYY";
+      String monthPattern = "MM";
+      String dayPattern = "DD";
+      String hourPattern = "HH";
+      String minutePattern = "mm";
+
+      // Use a custom pattern, if provided.
+      if (yearPattern() != null) {
+        yearPattern = MoreObjects.firstNonNull(yearPattern().get(), yearPattern);
+      }
+      LOG.debug("Year pattern set to: {}", yearPattern);
+
+      if (monthPattern() != null) {
+        monthPattern = MoreObjects.firstNonNull(monthPattern().get(), monthPattern);
+      }
+      LOG.debug("Month pattern set to: {}", monthPattern);
+
+      if (dayPattern() != null) {
+        dayPattern = MoreObjects.firstNonNull(dayPattern().get(), dayPattern);
+      }
+      LOG.debug("Day pattern set to: {}", dayPattern);
+
+      if (hourPattern() != null) {
+        hourPattern = MoreObjects.firstNonNull(hourPattern().get(), hourPattern);
+      }
+      LOG.debug("Hour pattern set to: {}", hourPattern);
+
+      if (minutePattern() != null) {
+        minutePattern = MoreObjects.firstNonNull(minutePattern().get(), minutePattern);
+      }
+      LOG.debug("Minute pattern set to: {}", minutePattern);
+
+      try {
+        final DateTimeFormatter yearFormatter = DateTimeFormat.forPattern(yearPattern);
+        final DateTimeFormatter monthFormatter = DateTimeFormat.forPattern(monthPattern);
+        final DateTimeFormatter dayFormatter = DateTimeFormat.forPattern(dayPattern);
+        final DateTimeFormatter hourFormatter = DateTimeFormat.forPattern(hourPattern);
+        final DateTimeFormatter minuteFormatter = DateTimeFormat.forPattern(minutePattern);
+    
+        outputPath = outputPath.replace(yearPattern, yearFormatter.print(time));
+        outputPath = outputPath.replace(monthPattern, monthFormatter.print(time));
+        outputPath = outputPath.replace(dayPattern, dayFormatter.print(time));
+        outputPath = outputPath.replace(hourPattern, hourFormatter.print(time));
+        outputPath = outputPath.replace(minutePattern, minuteFormatter.print(time));
+      } catch (IllegalArgumentException e) {
+        LOG.error("Invalid DateTime pattern: {}", e.getMessage());
+        throw new RuntimeException(e);
```

Comment:
```suggestion
        throw new RuntimeException("Invalid DateTime pattern", e);
```
Usually logging and re-throwing the same exception is considered an anti-pattern. Any strong reason to do it here?

Reply:
Not really. I saw it at a few other places in the repo back then. I'll can take out the log statements.

Reply:
Should I remove the try-catch block as well? The catch block was just to log the error message and throw a new RuntimeException. If I remove the log statement all we're doing is catching an IllegalArgumentException and throwing a RuntimeException.

Reply:
I'd keep it because IllegalArgumentException doesn't make it clear for the user what pattern is incorrect (e.g. if an empty pattern is supplied it just says "Invalid pattern specification"). Here the RuntimeException at least says that the error was caused by a DateTime pattern.

I'd go even further and include even more details in the error message, e.g. what pattern is incorrect (add a helper method to avoid try-catch around each of the 5 patterns). But it's up to you if you want to make such a change.

Reply:
I was wondering if it's a good idea to throw a RuntimeException... Since this resolution happens after we read from Pub/Sub, users may lose some data. Maybe we should use the default values if the pattern provided is invalid? WDYT?

Reply:
I'd still throw the Exception. If a user set a pattern override, this means the default pattern doesn't work for them for some reason. If the override was set incorrectly, simply ignoring this and falling back to the fault pattern may cause unwanted side effects (e.g. output written to a wrong folder, potentially overwriting existing data). IMO it's better to throw an exception and not process anything in this case.

As for the data loss, it shouldn't happen. Dataflow guarantees "exactly once" processing semantics ([some info about it](https://cloud.google.com/dataflow/docs/concepts/streaming-with-cloud-pubsub)). In case of PubSub, messages read from a topic are not going to be ACKed if there is a failure in the middle of the pipeline and PubSub is going to re-deliver them on the next processing attempt.

Reply:
Perfect. I'll leave the RuntimeException there then